server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -679,78 +670,99 @@ private void writeUpdatedShardMetadataAndComputeDeletes(SnapshotId snapshotId, R
 
         // Listener that flattens out the delete results for each index
         final ActionListener<Collection<ShardSnapshotMetaDeleteResult>> deleteIndexMetadataListener = new GroupedActionListener<>(
-            ActionListener.map(onAllShardsCompleted, res -> res.stream().flatMap(Collection::stream).collect(Collectors.toList())), indices.size());
+            ActionListener.map(onAllShardsCompleted,
+                               res -> res.stream().flatMap(Collection::stream).collect(Collectors.toList())),
+            indices.size());
 
         for (IndexId indexId : indices) {
             final Set<SnapshotId> survivingSnapshots = oldRepositoryData.getSnapshots(indexId).stream()
-                .filter(id -> id.equals(snapshotId) == false).collect(Collectors.toSet());
-            executor.execute(ActionRunnable.wrap(deleteIndexMetadataListener, deleteIdxMetaListener -> {
-                final StepListener<IndexMetadata> snapshotIndexMetadataListener = new StepListener<>();
-                try {
-                    getSnapshotIndexMetadata(snapshotId, indexId, snapshotIndexMetadataListener);
-                } catch (Exception ex) {
-                    LOGGER.warn(() -> new ParameterizedMessage("[{}] [{}] failed to read metadata for index", snapshotId, indexId.getName()), ex);
-                    // Just invoke the listener without any shard generations to count it down, this index will be cleaned up
-                    // by the stale data cleanup in the end.
-                    // TODO: Getting here means repository corruption. We should find a way of dealing with this instead of just ignoring
-                    //       it and letting the cleanup deal with it.
-                    deleteIdxMetaListener.onResponse(null);
+                .filter(id -> snapshotIds.contains(id) == false).collect(Collectors.toSet());
+            final StepListener<Collection<Integer>> shardCountListener = new StepListener<>();
+            final ActionListener<Integer> allShardCountsListener = new GroupedActionListener<>(shardCountListener,
+                                                                                               snapshotIds.size());
+            for (SnapshotId snapshotId : snapshotIds) {
+                executor.execute(ActionRunnable.wrap(allShardCountsListener, indexMetaDataListener -> {
```

Comment:
Please double check this part against: https://github.com/elastic/elasticsearch/commit/3a64ecb6bfe#diff-b9c15e79024348d03f0af119486a84c301b54135cc309c638834c8dfd9c101b0R676
since we are using a listener to get index metadata.

Please double check the whole method as well, thank you!